### PR TITLE
creating a vision subsystem

### PR DESCRIPTION
**Pull request by ShayKalisher1234**

### CODE REVIEW

**Comment by ori-coval** [comment]
the code isn't building fix it

File: src/main/java/frc/robot/Constants.java [hunk 1]
```diff
@@ -15,5 +30,20 @@
...
     public static final int kDriverControllerPort = 0;
+  }
+  public static class VisionConstants{
+    public static final String FRONT_CAMERA_NAME = "FRONT CAMERA NAME";
+    public static final AprilTagFieldLayout kTagLayout = AprilTagFields.kDefaultField.loadAprilTagLayoutField();
+    public static final Transform3d kRobotToCam = new Transform3d(new Translation3d(0.5, 0.0, 0.5), new Rotation3d(0, 0, 0));
+    public static final Matrix<N3, N1> kSingleTagStdDevs = VecBuilder.fill(4, 4, 8);
+    public static final Matrix<N3, N1> kMultiTagStdDevs = VecBuilder.fill(0.5, 0.5, 1);
+    //public static final Pose2d target = new Translation2d(1, 1); //the target
+    public static final Pose2d target = new Pose2d(1, 1, new Rotation2d(Units.degreesToRadians(0));
```
Comment (ori-coval) [inline src/main/java/frc/robot/Constants.java:41]: missing )

File: src/main/java/frc/robot/subsystems/Vision.java [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    private final PhotonCamera camera;
+    private final PhotonPoseEstimator photonEstimator;
+    double lastEstTimestamp = 0;
+    double distanceFromTarget;
+    double currentXPosition;
+    double currentYPosition;
+    double sideA;
+    double sideB;
+    double sideC;
+    double angle;
```
Comment (ori-coval) [inline src/main/java/frc/robot/subsystems/Vision.java:41]: all of this are useless remove them
`    double distanceFromTarget;
    double currentXPosition;
    double currentYPosition;
    double sideA;
    double sideB;
    double sideC;
    double angle;`

File: src/main/java/frc/robot/subsystems/Vision.java [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+
+    public Vision() {
```
Comment (ori-coval) [inline src/main/java/frc/robot/subsystems/Vision.java:43]: add singleton implemintation

**Review by ori-coval — CHANGES_REQUESTED** [review]

**Comment by danielml2** [comment]
> the code isn't building fix it

on that topic, maybe it's fine on your end, but according to the CI's log, there's a missing ) on [line 41 of constants](https://github.com/primo4586/Turret-learning-project/blob/TurretVision/src%2Fmain%2Fjava%2Ffrc%2Frobot%2FConstants.java#L41)

If this error doesn't show on your end you could move it down a line and commit and it should fix it
@ShayKalisher1234

**Comment by danielml2** [comment]
that's crazy i commented just as he submitted on the review but yeah

File: src/main/java/frc/robot/subsystems/Vision.java
Comment (ori-coval) [inline src/main/java/frc/robot/subsystems/Vision.java:1]: make all the functions static

File: src/main/java/frc/robot/subsystems/Vision.java [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+import edu.wpi.first.math.numbers.N3;
+import static frc.robot.Constants.VisionConstants.*;
+import frc.robot.Robot;
+
+
+
+public class Vision {
+    private AprilTagFieldLayout kTagLayout = AprilTagFields.k2023ChargedUp.loadAprilTagLayoutField();
+    private static PhotonCamera camera;
+    private static PhotonPoseEstimator photonEstimator;
```
Comment (ori-coval) [inline src/main/java/frc/robot/subsystems/Vision.java:35]: initialize variables here

**Review by ori-coval — CHANGES_REQUESTED** [review]

**Review by ori-coval — APPROVED** [review]

**Pull request merged**